File: Linked List/06 Palindrome_Linked_list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:

        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next

        prev = None

        while slow is not None:
            next_node = slow.next
            slow.next = prev
            prev = slow
            slow = next_node

        ptr = head
        second = prev

        while second is not None:
            if ptr.val != second.val:
                return False

            ptr = ptr.next
            second = second.next

        return True

File: Linked List/06 Palindrome_Linked_list/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val === undefined ? 0 : val)
 *     this.next = (next === undefined ? null : next)
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {

    // Find the middle of the list
    let slow = head;
    let fast = head;

    while (fast !== null && fast.next !== null) {
        slow = slow.next;
        fast = fast.next.next;
    }

    // Reverse the second half
    let prev = null;

    while (slow !== null) {
        let next = slow.next;
        slow.next = prev;
        prev = slow;
        slow = next;
    }

    // Compare first half with reversed second half
    let ptr = head;
    let second = prev;

    while (second !== null) {
        if (ptr.val !== second.val) {
            return false;
        }

        ptr = ptr.next;
        second = second.next;
    }

    return true;
};
